biz.aQute.bndlib/src/aQute/bnd/build/WorkspaceLock.java

```diff
@@ -103,6 +104,30 @@ private TimeoutException timeout(Lock lock) {
 		return e;
 	}
 
+	<T, U> T writeReadLocked(final long timeoutInMs, final Callable<U> underWrite,
+		final FunctionWithException<U, T> underRead, final BooleanSupplier canceled) throws Exception {
+		Callable<U> writeLocked = () -> {
+			U writeResult = underWrite.call();
+			// It is safe to let an exception propagate from here.
+			// Downgrade lock by taking read lock before returning to release
+			// write lock.
+			trace("Downgrading", writeLock());
+			readLock().lock();
+			return writeResult;
```

Comment:
Although I cannot see it can happen, it bugs me that an exception in the remainder of the return path would lose a read lock. I know it is not necessary looking at the code but I would call the `writeLocked` lambda in the `try/finally` that use for the `readLocked` and check a flag if we got/did not get the lock. I agree it is very defensive and not necessary :-)

Reply:
There is nothing left in the return path which would throw an exception and lose the read lock. I carefully checked this.